cilium/CFP-39876-clustermesh-filtered-export.md

```diff
@@ -0,0 +1,211 @@
+
+#### Allow/Deny listing namespace
+To allowlist a namespace and mark it as global, users can annotate the namespace with:
+
+```
+clustermesh.cilium.io/global: "true"
+```
+
+The ClusterMesh API server will watch namespace resources. Adding, changing or removing this namespace annotation will result in the following steps:
+
+- A full synchronization is triggered for all resources within the newly annotated namespaces 
+- Only entries associated with namespaces explicitly annotated as global will be exported.
+- Unannotated namespaces are treated as global/local depending on the value of config `clustermesh-default-global-namespace`
+
+To explicitly mark a namespace as local, the annotation can be set as:
+
+```
+clustermesh.cilium.io/global: "false"
+```
+
+When the annotation is removed or disabled for a namespace:
+- A full synchronization is triggered to add/remove the CiliumEndpoints and CiliumIdentities objects in etcd associated with the resources in the unannotated namespace
+- The system exports the CiliumEndpoints and CiliumIdentities with the updated list of global namespaces
+
+#### MCS Support 
+MCS-API ServiceExport and ServiceImport CRs will be only supported from a global Namespace. On a local namespace the following will be happening:
+- ServiceExport created by a user will report a Condition to signal that the namespace is local and the Service is not actually exported to the clustermesh-apiserver.
+- ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.
+
+### ClusterMesh API Server Changes
+
+#### Process annotations 
+We will add new watch on namespaces to check for global or local annotations and maintain internal sets for the same. Once we have the list, we will trigger a full sync operation to update the export based on the provided annotation. 
+
+#### Export Filtering
+Only the CiliumIdentities and CiliumEndpoints under global namespaces will be exported to the remote clusters. 
+
+#### ServiceExport controller 
+Provide warning message when a ServiceExport is triggered for a service residing in local namespace 
+
+#### Existing Cilium Resource Types
+- CiliumEndpoint(Slice)
+- CiliumIdentities
+- Global Services
+- MCS ServiceExport
+
+## Impacts / Key Questions
+---
+
+### Impact: Breakage during upgrade
+- The changes are expected to be backward compatible in the sense that the new scoped export logic only takes effect once the user has explicitly annotated namespaces as global or local. As a result, existing clusters will continue to function as before until the user opts into the new behavior. The `clustermesh-default-global-namespace` config option will determine the default behavior for unannotated namespaces.
```

Comment:
today, all identities are shared across all clusters. When user upgrades to this new version, that means upgrading to new agent and apiserver, existing workloads with network policies across clusters continue to work without annotating anything? and only new deployments requires annotation?

Reply:
The annotation is not required in any case. It acts as a feature flag for users to opt into this scoped-export mode. Without annotation, the behavior of ClusterMesh would be exactly the same as it is today